File: R/structIntervDist.R
#llm
library(methods)
library(igraph)

source("computePathMatrix.R")
source("computePathMatrix2.R")
source("allDagsJonas.R")
source("dSepAdji.R")

structIntervDist <- function(trueGraph, estGraph, output = FALSE, spars = FALSE)
    # Copyright (c) 2013 - 2013  Jonas Peters  [peters@stat.math.ethz.ch]
    # All rights reserved.  See the file COPYING for license terms.
{
    # These are all initializations
    estGraph <- as(estGraph, "matrix")
    trueGraph <- as(trueGraph, "matrix")
    p <- dim(trueGraph)[2]
    rownames(estGraph) <- 1:p
    colnames(estGraph) <- 1:p

    incorrectInt <- matrix(0, p, p)
    correctInt <- matrix(0, p, p)
    minimumTotal <- 0
    maximumTotal <- 0
    timePathMatrix2 <- 0
    timeAllComputePM2 <- 0
    timeAllComputePM <- 0
    timeAlldSep <- 0
    timeExpGraph <- 0
    numChecks <- 0
    Gtmp <- diag(1, p)
    ptmtotal <- proc.time()

    # Compute the path matrix whose entry (i,j) is TRUE if there is a directed path
    # from i to j. The diagonal is TRUE, too.
    PathMatrix <- computePathMatrix(trueGraph, spars)
    PathMatrix <- as(PathMatrix, "matrix")

    # Undirected part of estimated graph
    Gp.undir <- estGraph * t(estGraph)

    # Build undirected graph directly from adjacency matrix
    g.undir <- igraph::graph_from_adjacency_matrix(
        (Gp.undir > 0) * 1,
        mode = "undirected",
        diag = FALSE
    )

    comp.obj <- igraph::components(g.undir)
    conn.comp <- split(seq_len(p), comp.obj$membership)
    numConnComp <- length(conn.comp)

    GpIsEssentialGraph <- TRUE

    for (ll in 1:numConnComp)
    {
        conn.comp[[ll]] <- as.numeric(conn.comp[[ll]])

        if (length(conn.comp[[ll]]) > 1)
        {
            sub_undir <- (Gp.undir[conn.comp[[ll]], conn.comp[[ll]]] > 0) * 1

            sub_graph <- igraph::graph_from_adjacency_matrix(
                sub_undir,
                mode = "undirected",
                diag = FALSE
            )

            chordal <- igraph::is.chordal(sub_graph, fillin = TRUE)$chordal

            if (!chordal)
            {
                show("The estimated graph is not chordal, i.e. it is not a CPDAG! We thus consider local expansions of the graph (some combinations of which may lead to cycles).")
                GpIsEssentialGraph <- FALSE
            }

            if (length(conn.comp[[ll]]) > 8)
            {
                show("The connected component is too large (>8 nodes) in order to be extended to all DAGs in a reasonable amount of time. We thus consider local expansions of the graph (some combinations of which may lead to cycles).")
                GpIsEssentialGraph <- FALSE
            }
        }
    }

    for (ll in 1:numConnComp)
    {
        ptm <- proc.time()

        if (length(conn.comp[[ll]]) > 0)
        {
            if (GpIsEssentialGraph)
            {
                # expand the connected component into DAGs
                if (length(conn.comp[[ll]]) > 1)
                {
                    mmm <- allDagsJonas(estGraph, conn.comp[[ll]])
                }
                else
                {
                    mmm <- matrix(estGraph, 1, p^2)
                }

                if (sum(mmm == -1) == 1)
                {
                    GpIsEssentialGraph <- FALSE
                    mmm <- matrix(estGraph, 1, p^2)
                }

                # each row in mmm contains one DAG.
                # the first p entries of estGraph are the parents of node 1
                # in the following lines we change this to: the first p entries of estGraph are the children of node 1
                newInd <- seq(1, p^3, by = p) -
                    rep(seq(0, (p - 1) * (p^2 - 1), by = (p^2 - 1)), each = p)

                dimM <- dim(mmm)
                mmm <- matrix(mmm[, newInd], dimM)

                if (is.null(mmm))
                {
                    show("Something is wrong. Maybe the estimated graph is not a CPDAG? We expand the undirected components locally.")
                    GpIsEssentialGraph <- FALSE
                }
                else
                {
                    incorrectSum <- rep(0, dim(mmm)[1])
                }
            }
        }

        timeExpGraph <- timeExpGraph + proc.time() - ptm

        for (i in conn.comp[[ll]])
        {
            paG <- which(trueGraph[, i] == 1)
            certainpaGp <- which((estGraph[, i] * (rep(1, p) - estGraph[i, ])) == 1)
            possiblepaGp <- which((estGraph[, i] * estGraph[i, ]) == 1)

            if (!GpIsEssentialGraph)
            {
                maxcount <- 2^length(possiblepaGp)
                uniqueRows <- 1:maxcount
                mmm <- t(matrix(rep(t(estGraph)[1:length(estGraph)], maxcount), length(estGraph), maxcount))
                mmm[, i + (possiblepaGp - 1) * p] <- as.matrix(
                    expand.grid(rep(list(0:1), length(possiblepaGp)))
                )
                incorrectSum <- rep(0, maxcount)
            }
            else
            {
                if (dim(mmm)[1] > 1)
                {
                    allParentsOfI <- seq(i, (p - 1) * p + i, by = p)
                    uniqueRows <- which(!duplicated(mmm[, allParentsOfI]))
                    maxcount <- length(uniqueRows)
                }
                else
                {
                    maxcount <- 1
                    uniqueRows <- 1
                }
            }

            count <- 1

            while (count <= maxcount)
            {
                if (maxcount == 1)
                {
                    paGp <- certainpaGp
                }
                else
                {
                    Gpnew <- t(matrix(mmm[uniqueRows[count], ], p, p))
                    paGp <- which(Gpnew[, i] == 1)

                    if (output)
                    {
                        cat(i, " has ", length(paGp), " parents in expansion nr. ", uniqueRows[count], " of Gp:")
                        show(paGp)
                    }
                }

                # the following computations are the same for all j (i is fixed)
                ptm <- proc.time()
                PathMatrix2 <- computePathMatrix2(trueGraph, paGp, PathMatrix, spars)
                timePathMatrix2 <- timePathMatrix2 + proc.time() - ptm

                ptm <- proc.time()
                checkAlldSep <- dSepAdji(trueGraph, i, paGp, PathMatrix, PathMatrix2, spars = spars)
                numChecks <- numChecks + 1
                timeAlldSep <- timeAlldSep + proc.time() - ptm
                timeAllComputePM2 <- timeAllComputePM2 + checkAlldSep$timeComputePM2
                timeAllComputePM <- timeAllComputePM + checkAlldSep$timeComputePM
                reachableWOutCausalPath <- checkAlldSep$reachableOnNonCausalPath

                for (j in 1:p)
                {
                    if (i != j)
                    {
                        # The order of the following checks and the flag finished are
                        # made such that as few tests are performed as possible.

                        finished <- FALSE
                        ijGNull <- FALSE
                        ijGpNull <- FALSE

                        # ijGNull means that the causal effect from i to j is zero in G
                        # more precisely, p(x_j | do (x_i=a)) = p(x_j)
                        if (PathMatrix[i, j] == 0)
                        {
                            ijGNull <- TRUE
                        }

                        # if j->i exists in Gp
                        if ((sum(paGp == j) == 1))
                        {
                            ijGpNull <- TRUE
                        }

                        # if both are zero
                        if (ijGpNull & ijGNull)
                        {
                            finished <- TRUE
                            correctInt[i, j] <- 1
                        }

                        # if Gp predicts zero but G says it is not
                        if (ijGpNull & !ijGNull)
                        {
                            incorrectInt[i, j] <- 1
                            incorrectSum[uniqueRows[count]] <- incorrectSum[uniqueRows[count]] + 1

                            ###############
                            # also add one to all entries of incorrectSum that have the same set of parents of i - find them
                            allOthers <- setdiff(1:(dim(mmm)[1]), uniqueRows)

                            if (length(allOthers) > 1)
                            {
                                indInAllOthers <- which(
                                    colSums(!xor(mmm[uniqueRows[count], allParentsOfI], t(mmm[allOthers, allParentsOfI]))) == p
                                )
                                if (length(indInAllOthers) > 0)
                                {
                                    incorrectSum[allOthers[indInAllOthers]] <- incorrectSum[allOthers[indInAllOthers]] + rep(1, length(indInAllOthers))
                                }
                            }

                            if (length(allOthers) == 1)
                            {
                                indInAllOthers <- which(
                                    sum(!xor(mmm[uniqueRows[count], allParentsOfI], t(mmm[allOthers, allParentsOfI]))) == p
                                )
                                if (length(indInAllOthers) > 0)
                                {
                                    incorrectSum[allOthers[indInAllOthers]] <- incorrectSum[allOthers[indInAllOthers]] + rep(1, length(indInAllOthers))
                                }
                            }
                            ###############

                            finished <- TRUE
                        }

                        # if the set of parents are the same
                        if (!finished && setequal(paG, paGp))
                        {
                            finished <- TRUE
                            correctInt[i, j] <- 1
                        }

                        # this part contains the difficult computations
                        if (!finished)
                        {
                            if (PathMatrix[i, j] > 0)
                            {
                                # which children are part of a causal path?
                                chiCausPath <- which(trueGraph[i, ] & PathMatrix[, j])

                                # check whether in paGp there is a descendant of a "proper" child of i
                                if (sum(PathMatrix[chiCausPath, paGp]) > 0)
                                {
                                    incorrectInt[i, j] <- 1
                                    incorrectSum[uniqueRows[count]] <- incorrectSum[uniqueRows[count]] + 1

                                    ###############
                                    allOthers <- setdiff(1:(dim(mmm)[1]), uniqueRows)

                                    if (length(allOthers) > 1)
                                    {
                                        indInAllOthers <- which(
                                            colSums(!xor(mmm[uniqueRows[count], allParentsOfI], t(mmm[allOthers, allParentsOfI]))) == p
                                        )
                                        if (length(indInAllOthers) > 0)
                                        {
                                            incorrectSum[allOthers[indInAllOthers]] <- incorrectSum[allOthers[indInAllOthers]] + rep(1, length(indInAllOthers))
                                        }
                                    }

                                    if (length(allOthers) == 1)
                                    {
                                        indInAllOthers <- which(
                                            sum(!xor(mmm[uniqueRows[count], allParentsOfI], t(mmm[allOthers, allParentsOfI]))) == p
                                        )
                                        if (length(indInAllOthers) > 0)
                                        {
                                            incorrectSum[allOthers[indInAllOthers]] <- incorrectSum[allOthers[indInAllOthers]] + rep(1, length(indInAllOthers))
                                        }
                                    }
                                    ###############

                                    finished <- TRUE
                                }
                            }

                            if (!finished)
                            {
                                # check whether all non-causal paths are blocked
                                if (reachableWOutCausalPath[j] == 1)
                                {
                                    incorrectInt[i, j] <- 1
                                    incorrectSum[uniqueRows[count]] <- incorrectSum[uniqueRows[count]] + 1

                                    ###############
                                    allOthers <- setdiff(1:(dim(mmm)[1]), uniqueRows)

                                    if (length(allOthers) > 1)
                                    {
                                        indInAllOthers <- which(
                                            colSums(!xor(mmm[uniqueRows[count], allParentsOfI], t(mmm[allOthers, allParentsOfI]))) == p
                                        )
                                        if (length(indInAllOthers) > 0)
                                        {
                                            incorrectSum[allOthers[indInAllOthers]] <- incorrectSum[allOthers[indInAllOthers]] + rep(1, length(indInAllOthers))
                                        }
                                    }

                                    if (length(allOthers) == 1)
                                    {
                                        indInAllOthers <- which(
                                            sum(!xor(mmm[uniqueRows[count], allParentsOfI], t(mmm[allOthers, allParentsOfI]))) == p
                                        )
                                        if (length(indInAllOthers) > 0)
                                        {
                                            incorrectSum[allOthers[indInAllOthers]] <- incorrectSum[allOthers[indInAllOthers]] + rep(1, length(indInAllOthers))
                                        }
                                    }
                                    ###############
                                }
                                else
                                {
                                    correctInt[i, j] <- 1
                                }
                            }
                        }
                    }
                } # for-loop over j

                count <- count + 1
            } # while-loop count<=maxcount

            if (!GpIsEssentialGraph)
            {
                minimumTotal <- minimumTotal + min(incorrectSum)
                maximumTotal <- maximumTotal + max(incorrectSum)
                incorrectSum <- 0
            }

            if (length(incorrectSum) > 1 & output)
            {
                cat("For variable ", i, "we have more than one possible set of parents.\n")
                cat("The following vector is adding the number of incorrect interventions for the possible parent sets.\n")
                cat("(the computation is done in a clever way since some of the components may correspond to the same parent set of node ", i, ").\n")
                show(incorrectSum)
                cat("Only for a new connected component this vector is set to zero again.\n\n")
            }
        } # i in conn.comp

        minimumTotal <- minimumTotal + min(incorrectSum)
        maximumTotal <- maximumTotal + max(incorrectSum)
        incorrectSum <- 0
    }

    timeTotal <- proc.time() - ptmtotal

    # The rest is output and return.
    if (output && p < 11)
    {
        show("These all are incorrectly predicted interventions:")
        show(incorrectInt)
        show("And these are all correctly predicted interventions:")
        show(correctInt)
    }

    ress <- list()
    ress$sid <- sum(incorrectInt)
    ress$sidUpperBound <- maximumTotal
    ress$sidLowerBound <- minimumTotal
    ress$incorrectMat <- incorrectInt

    if (output)
    {
        cat("Time needed for ... \n")
        cat("... expanding the graph: ", timeExpGraph[3], "\n")
        cat("... computing path matrices (used for checking d-seps): ", timePathMatrix2[3], "\n")
        cat("... checking d-separations: ", timeAlldSep[3], "\n")
        cat("... ... herof: computePathMatrix2: ", timeAllComputePM2, "\n")
        cat("... ... herof: computePathMatrix: ", timeAllComputePM, "\n")
        cat("... in total: ", timeTotal[3], "\n")
        cat("number of times we ran *check all d-seps*:", numChecks, "\n")
    }

    return(ress)
}


read_testcase <- function(filepath)
{
    lines <- readLines(filepath, warn = FALSE)

    trueGraph <- list()
    estGraph <- list()
    output <- FALSE
    spars <- FALSE
    mode <- NULL

    for (line in lines)
    {
        line <- trimws(line)

        if (line == "")
        {
            next
        }

        if (line == "trueGraph:")
        {
            mode <- "trueGraph"
            next
        }

        if (line == "estGraph:")
        {
            mode <- "estGraph"
            next
        }

        if (line == "output:")
        {
            mode <- "output"
            next
        }

        if (line == "spars:")
        {
            mode <- "spars"
            next
        }

        if (mode == "trueGraph")
        {
            trueGraph[[length(trueGraph) + 1]] <- as.integer(strsplit(line, " +")[[1]])
        }
        else if (mode == "estGraph")
        {
            estGraph[[length(estGraph) + 1]] <- as.integer(strsplit(line, " +")[[1]])
        }
        else if (mode == "output")
        {
            output <- tolower(line) == "true"
        }
        else if (mode == "spars")
        {
            spars <- tolower(line) == "true"
        }
    }

    trueGraph <- do.call(rbind, trueGraph)
    estGraph <- do.call(rbind, estGraph)

    return(list(
        trueGraph = trueGraph,
        estGraph = estGraph,
        output = output,
        spars = spars
    ))
}


script_dir <- getwd()
project_dir <- dirname(script_dir)

testcase_dir <- file.path(project_dir, "tests", "structIntervDist", "Testcase")
output_dir <- file.path(project_dir, "tests", "structIntervDist", "R_outputs")

if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
}

output_file <- file.path(output_dir, "all_results.txt")
out <- file(output_file, open = "w")

for (i in 1:100)
{
    filepath <- file.path(testcase_dir, paste0(i, ".txt"))
    values <- read_testcase(filepath)

    trueGraph <- values$trueGraph
    estGraph <- values$estGraph
    output_flag <- values$output
    spars_flag <- values$spars

    result <- structIntervDist(
        trueGraph,
        estGraph,
        output = output_flag,
        spars = spars_flag
    )

    writeLines(paste0("Testcase ", i), out)

    writeLines("trueGraph:", out)
    for (r in 1:nrow(trueGraph)) {
        writeLines(paste(trueGraph[r, ], collapse = " "), out)
    }

    writeLines("estGraph:", out)
    for (r in 1:nrow(estGraph)) {
        writeLines(paste(estGraph[r, ], collapse = " "), out)
    }

    writeLines("output:", out)
    writeLines(as.character(output_flag), out)

    writeLines("spars:", out)
    writeLines(as.character(spars_flag), out)

    writeLines("sid:", out)
    writeLines(as.character(result$sid), out)

    writeLines("sidUpperBound:", out)
    writeLines(as.character(result$sidUpperBound), out)

    writeLines("sidLowerBound:", out)
    writeLines(as.character(result$sidLowerBound), out)

    writeLines("incorrectMat:", out)
    for (r in 1:nrow(result$incorrectMat)) {
        writeLines(paste(result$incorrectMat[r, ], collapse = " "), out)
    }

    writeLines("", out)
}

close(out)

File: py_impl/structIntervDist.py

import os
import time
import itertools
import numpy as np
import networkx as nx

from computePathMatrix import compute_path_matrix
from computePathMatrix2 import compute_path_matrix2
from allDagsJonas import all_dags_jonas
from dSepAdji import d_sep_adji
#llm

def _same_parent_columns(i, p):
    return (i - 1) + np.arange(p) * p


def _add_incorrect_to_duplicate_parent_sets(incorrect_sum, mmm, current_row_idx, unique_rows, parent_cols):
    all_rows = np.arange(mmm.shape[0])
    unique_row_set = set(unique_rows)
    current_parent_pattern = mmm[current_row_idx, parent_cols]

    for row_idx in all_rows:
        if row_idx in unique_row_set:
            continue
        if np.array_equal(current_parent_pattern, mmm[row_idx, parent_cols]):
            incorrect_sum[row_idx] += 1


def struct_interv_dist(true_graph, est_graph, output=False, spars=False):
    est_graph = np.array(est_graph, dtype=int)
    true_graph = np.array(true_graph, dtype=int)

    p = true_graph.shape[1]

    incorrect_int = np.zeros((p, p), dtype=int)
    correct_int = np.zeros((p, p), dtype=int)

    minimum_total = 0
    maximum_total = 0

    time_path_matrix2 = 0.0
    time_all_compute_pm2 = 0.0
    time_all_compute_pm = 0.0
    time_all_d_sep = 0.0
    time_exp_graph = 0.0

    num_checks = 0
    gtmp = np.eye(p, dtype=int)
    ptm_total = time.time()

    path_matrix = np.array(compute_path_matrix(true_graph, spars=spars), dtype=int)

    gp_undir = est_graph * est_graph.T
    undir_graph = nx.from_numpy_array((gp_undir > 0).astype(int))

    conn_comp = [
        np.array(sorted([node + 1 for node in comp]), dtype=int)
        for comp in nx.connected_components(undir_graph)
    ]
    num_conn_comp = len(conn_comp)

    gp_is_essential_graph = True

    for ll in range(num_conn_comp):
        comp = conn_comp[ll]

        if len(comp) > 1:
            sub = gp_undir[np.ix_(comp - 1, comp - 1)]
            sub_graph = nx.from_numpy_array((sub > 0).astype(int))
            chordal = nx.is_chordal(sub_graph)

            if not chordal:
                print(
                    "The estimated graph is not chordal, i.e. it is not a CPDAG! "
                    "We thus consider local expansions of the graph "
                    "(some combinations of which may lead to cycles)."
                )
                gp_is_essential_graph = False

            if len(comp) > 8:
                print(
                    "The connected component is too large (>8 nodes) in order to be "
                    "extended to all DAGs in a reasonable amount of time. "
                    "We thus consider local expansions of the graph "
                    "(some combinations of which may lead to cycles)."
                )
                gp_is_essential_graph = False

    for ll in range(num_conn_comp):
        ptm = time.time()
        comp = conn_comp[ll]

        if len(comp) > 0:
            if gp_is_essential_graph:
                if len(comp) > 1:
                    mmm = all_dags_jonas(est_graph, comp)
                    mmm = np.array(mmm, dtype=int)
                    if mmm.ndim == 1:
                        mmm = mmm.reshape(1, -1)
                else:
                    mmm = est_graph.flatten(order="F").reshape(1, p**2)

                if np.sum(mmm == -1) == 1:
                    gp_is_essential_graph = False
                    mmm = est_graph.flatten(order="F").reshape(1, p**2)

                # Convert each DAG row to the same row-wise layout used later in the checks
                mmm = np.array(
                    [row.reshape((p, p), order="F").flatten(order="C") for row in mmm],
                    dtype=int
                )

                if mmm is None:
                    print(
                        "Something is wrong. Maybe the estimated graph is not a CPDAG? "
                        "We expand the undirected components locally."
                    )
                    gp_is_essential_graph = False
                else:
                    incorrect_sum = np.zeros(mmm.shape[0], dtype=int)

        time_exp_graph += time.time() - ptm

        for i in comp:
            pa_g = np.where(true_graph[:, i - 1] == 1)[0] + 1
            certain_pa_gp = np.where(
                (est_graph[:, i - 1] * (np.ones(p, dtype=int) - est_graph[i - 1, :])) == 1
            )[0] + 1
            possible_pa_gp = np.where(
                (est_graph[:, i - 1] * est_graph[i - 1, :]) == 1
            )[0] + 1

            all_parents_of_i = _same_parent_columns(i, p)

            if not gp_is_essential_graph:
                maxcount = 2 ** len(possible_pa_gp)
                unique_rows = np.arange(maxcount, dtype=int)

                base_row = est_graph.flatten(order="C")
                mmm = np.tile(base_row, (maxcount, 1))

                if len(possible_pa_gp) > 0:
                    combinations = np.array(
                        list(itertools.product([0, 1], repeat=len(possible_pa_gp))),
                        dtype=int
                    )
                    target_cols = (i - 1) + (possible_pa_gp - 1) * p
                    mmm[:, target_cols] = combinations

                incorrect_sum = np.zeros(maxcount, dtype=int)
            else:
                if mmm.shape[0] > 1:
                    seen = set()
                    unique_rows = []
                    for row_idx in range(mmm.shape[0]):
                        key = tuple(mmm[row_idx, all_parents_of_i].tolist())
                        if key not in seen:
                            seen.add(key)
                            unique_rows.append(row_idx)
                    unique_rows = np.array(unique_rows, dtype=int)
                    maxcount = len(unique_rows)
                else:
                    maxcount = 1
                    unique_rows = np.array([0], dtype=int)

            count = 0
            while count < maxcount:
                current_row_idx = unique_rows[count]

                if maxcount == 1:
                    pa_gp = certain_pa_gp.copy()
                else:
                    gp_new = mmm[current_row_idx, :].reshape((p, p), order="C")
                    pa_gp = np.where(gp_new[:, i - 1] == 1)[0] + 1

                    if output:
                        print(
                            i,
                            " has ",
                            len(pa_gp),
                            " parents in expansion nr. ",
                            current_row_idx + 1,
                            " of Gp:",
                            sep=""
                        )
                        print(pa_gp)

                ptm = time.time()
                path_matrix2 = compute_path_matrix2(
                    true_graph,
                    pa_gp.tolist(),
                    path_matrix,
                    spars=spars
                )
                time_path_matrix2 += time.time() - ptm

                ptm = time.time()
                check_all_d_sep = d_sep_adji(
                    true_graph,
                    i,
                    pa_gp.tolist(),
                    path_matrix=path_matrix,
                    path_matrix2=path_matrix2,
                    spars=spars
                )
                num_checks += 1
                time_all_d_sep += time.time() - ptm
                time_all_compute_pm2 += check_all_d_sep["timeComputePM2"]
                time_all_compute_pm += check_all_d_sep["timeComputePM"]

                reachable_without_causal_path = np.array(
                    check_all_d_sep["reachableOnNonCausalPath"],
                    dtype=int
                )

                for j in range(1, p + 1):
                    if i != j:
                        finished = False
                        ij_g_null = False
                        ij_gp_null = False

                        if path_matrix[i - 1, j - 1] == 0:
                            ij_g_null = True

                        if np.sum(pa_gp == j) == 1:
                            ij_gp_null = True

                        if ij_gp_null and ij_g_null:
                            finished = True
                            correct_int[i - 1, j - 1] = 1

                        if ij_gp_null and not ij_g_null:
                            incorrect_int[i - 1, j - 1] = 1
                            incorrect_sum[current_row_idx] += 1
                            _add_incorrect_to_duplicate_parent_sets(
                                incorrect_sum, mmm, current_row_idx, unique_rows, all_parents_of_i
                            )
                            finished = True

                        if (not finished) and set(pa_g.tolist()) == set(pa_gp.tolist()):
                            finished = True
                            correct_int[i - 1, j - 1] = 1

                        if not finished:
                            if path_matrix[i - 1, j - 1] > 0:
                                chi_caus_path = np.where(
                                    (true_graph[i - 1, :] > 0) & (path_matrix[:, j - 1] > 0)
                                )[0] + 1

                                if len(chi_caus_path) > 0 and len(pa_gp) > 0:
                                    if np.sum(path_matrix[np.ix_(chi_caus_path - 1, pa_gp - 1)]) > 0:
                                        incorrect_int[i - 1, j - 1] = 1
                                        incorrect_sum[current_row_idx] += 1
                                        _add_incorrect_to_duplicate_parent_sets(
                                            incorrect_sum, mmm, current_row_idx, unique_rows, all_parents_of_i
                                        )
                                        finished = True

                            if not finished:
                                if reachable_without_causal_path[j - 1] == 1:
                                    incorrect_int[i - 1, j - 1] = 1
                                    incorrect_sum[current_row_idx] += 1
                                    _add_incorrect_to_duplicate_parent_sets(
                                        incorrect_sum, mmm, current_row_idx, unique_rows, all_parents_of_i
                                    )
                                else:
                                    correct_int[i - 1, j - 1] = 1

                count += 1

            if not gp_is_essential_graph:
                minimum_total += int(np.min(incorrect_sum))
                maximum_total += int(np.max(incorrect_sum))
                incorrect_sum = np.array([0], dtype=int)

            if len(incorrect_sum) > 1 and output:
                print(f"For variable {i} we have more than one possible set of parents.")
                print("The following vector is adding the number of incorrect interventions for the possible parent sets.")
                print(
                    f"(the computation is done in a clever way since some of the components "
                    f"may correspond to the same parent set of node {i})."
                )
                print(incorrect_sum)
                print("Only for a new connected component this vector is set to zero again.\n")

        minimum_total += int(np.min(incorrect_sum))
        maximum_total += int(np.max(incorrect_sum))
        incorrect_sum = np.array([0], dtype=int)

    time_total = time.time() - ptm_total

    if output and p < 11:
        print("These all are incorrectly predicted interventions:")
        print(incorrect_int)
        print("And these are all correctly predicted interventions:")
        print(correct_int)

    result = {
        "sid": int(np.sum(incorrect_int)),
        "sidUpperBound": int(maximum_total),
        "sidLowerBound": int(minimum_total),
        "incorrectMat": incorrect_int
    }

    if output:
        print("Time needed for ...")
        print("... expanding the graph:", time_exp_graph)
        print("... computing path matrices (used for checking d-seps):", time_path_matrix2)
        print("... checking d-separations:", time_all_d_sep)
        print("... ... hereof: computePathMatrix2:", time_all_compute_pm2)
        print("... ... hereof: computePathMatrix:", time_all_compute_pm)
        print("... in total:", time_total)
        print("number of times we ran *check all d-seps*:", num_checks)

    return result


def read_testcase(filepath):
    true_graph = []
    est_graph = []
    output = False
    spars = False
    mode = None

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()

            if line == "":
                continue

            if line == "trueGraph:":
                mode = "trueGraph"
                continue

            if line == "estGraph:":
                mode = "estGraph"
                continue

            if line == "output:":
                mode = "output"
                continue

            if line == "spars:":
                mode = "spars"
                continue

            if mode == "trueGraph":
                true_graph.append([int(x) for x in line.split()])
            elif mode == "estGraph":
                est_graph.append([int(x) for x in line.split()])
            elif mode == "output":
                output = line.lower() == "true"
            elif mode == "spars":
                spars = line.lower() == "true"

    return (
        np.array(true_graph, dtype=int),
        np.array(est_graph, dtype=int),
        output,
        spars
    )


script_dir = os.getcwd()
project_dir = os.path.dirname(script_dir)

testcase_dir = os.path.join(project_dir, "tests", "structIntervDist", "Testcase")
output_dir = os.path.join(project_dir, "tests", "structIntervDist", "Py_outputs")

if not os.path.exists(output_dir):
    os.mkdir(output_dir)

output_file = os.path.join(output_dir, "all_results.txt")

print("Reading testcases from:", testcase_dir)
print("Saving output to:", output_file)

with open(output_file, "w", encoding="utf-8") as f:
    for i in range(1, 101):
        filepath = os.path.join(testcase_dir, f"{i}.txt")
        true_graph, est_graph, output_flag, spars_flag = read_testcase(filepath)

        result = struct_interv_dist(
            true_graph,
            est_graph,
            output=output_flag,
            spars=spars_flag
        )

        f.write(f"Testcase {i}\n")

        f.write("trueGraph:\n")
        for row in true_graph:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("estGraph:\n")
        for row in est_graph:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("output:\n")
        f.write(str(output_flag) + "\n")

        f.write("spars:\n")
        f.write(str(spars_flag) + "\n")

        f.write("sid:\n")
        f.write(str(result["sid"]) + "\n")

        f.write("sidUpperBound:\n")
        f.write(str(result["sidUpperBound"]) + "\n")

        f.write("sidLowerBound:\n")
        f.write(str(result["sidLowerBound"]) + "\n")

        f.write("incorrectMat:\n")
        for row in result["incorrectMat"]:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("\n")

print("Done. Output saved to:", output_file)

